library/swift/src/StreamClientBuilder.swift

```diff
@@ -107,6 +108,28 @@ public final class StreamClientBuilder: NSObject {
     return self
   }
 
+  /// Add HTTP filter for requests sent by this client.
+  ///
+  /// - parameter filter: RequestFilter to be invoked for streams.
+  ///
+  /// - returns: This builder.
+  @discardableResult
+  public func addFilter(_ filter: RequestFilter) -> StreamClientBuilder {
+    self.filterChain.append(EnvoyHTTPFilter(requestFilter: filter))
+    return self
+  }
+
+  /// Add HTTP filter for resonses received by this client.
+  ///
+  /// - parameter filter: RequestFilter to be invoked for streams.
+  ///
+  /// - returns: This builder.
+  @discardableResult
+  public func addFilter(_ filter: ResponseFilter) -> StreamClientBuilder {
+    self.filterChain.append(EnvoyHTTPFilter(responseFilter: filter))
+    return self
+  }
```

Comment:
1. It looks like this prevents filters from being able to act as both `RequestFilter` and `ResponseFilter` conformers. The idea behind the empty `Filter` protocol in place is to allow registration of a filter, then call its request/response functionality depending on what it conforms to. Should we be doing something like that here?
2. Per comments above, I think these should accept closures that create the filters on each request/stream, rather than instances of the filters.

Reply:
1. Agreed, see my response above. With a little bit of reflection I think we can achieve that, so I'm open to working that out here, or in a subsequent PR, though my slight preference is for the latter.
2. I agree, though I have to think a little bit more about the ramifications to the API and mapping layer. I'm forming an idea of how to do it though - it should be possible.

Reply:
re: 2. we should definitely do this per stream instance. And re: 1. do we really need reflection? Can't we have multiple inheritance and have a filter class that is both, and thus calls both request and response addFilter functions as in envoy: 

https://github.com/envoyproxy/envoy/blob/bc02b7748a69fd3854f74b9e5bec236b5499f7fd/include/envoy/http/filter.h#L772

https://github.com/envoyproxy/envoy/blob/0c6523c0925d260de41d89002044d0d6271550d0/source/common/http/conn_manager_impl.h#L553

Reply:
I believe reflection is the way to go, yes. It's true a class can implement multiple interfaces, but even if we declare an interface that conforms to both `RequestFilter` and `ResponseFilter`, Swift will not implicitly or automatically cause classes that implement both interfaces to conform to it. So when we go to instantiate bridge type from a generic `Filter`, we will need to check which interfaces it implements. If we create type-specialized constructors (like I had here), it will be ambiguous to the compiler which it should call.

That's the best explanation I can provide here. Happy to discuss more offline.

Reply:
For bi-directionality, see the solution I've pushed, and let me know if you have additional thoughts. :)